packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,129 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { EventEmitter } from "events";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { IConnectionDetails } from "@fluidframework/container-definitions";
+import { ProtocolOpHandler, Quorum } from "@fluidframework/protocol-base";
+import { ConnectionMode } from "@fluidframework/protocol-definitions";
+import { connectEventName, ConnectionState, ILocalSequencedClient } from "./container";
+
+export class ConnectionStateHandler {
+    private _connectionState = ConnectionState.Disconnected;
+    private _pendingClientId: string | undefined;
+    private _clientId: string | undefined;
+
+    public get connectionState(): ConnectionState {
+        return this._connectionState;
+    }
+
+    public get connected(): boolean {
+        return this.connectionState === ConnectionState.Connected;
+    }
+
+    public get clientId(): string | undefined {
+        return this._clientId;
+    }
+
+    public get pendingClientId(): string | undefined {
+        return this._pendingClientId;
+    }
+
+    constructor(
+        private readonly protocolHandler: () => ProtocolOpHandler | undefined,
+        private readonly logConnectionStateChangeTelemetry:
+            (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void,
+        private readonly propagateConnectionState: () => void,
+        private readonly isContainerLoaded: () => boolean,
```

Comment:
Nit: I think you can get all the callbacks into something like that:
```
interface INeedThis {
  readonly loaded: boolean;
  readonly protocolHandler: ProtocolOpHandler  | undefined;
...
}
```

And then let container implement it.
And maybe (I'm not sure) Container can just pass `this` as an implementation (not sure if it will work with protected / private fields - I think it will).

That way it will be easier to add or remove things in the future, and it's a bit more compact in usage.

Reply:
I made an interface but did not made container implement it as these methods are not used within the container at all.